pkg/controller/elasticsearch/driver/upgrade_deletion_strategy.go

```diff
@@ -0,0 +1,263 @@
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			if candidate.DeletionTimestamp != nil && candidate.Status.Reason == NodeUnreachablePodReason {
+				// kubelet is unresponsive, Unknown Pod, do not try to delete it
+				return false, nil
+			}
+			if candidate.DeletionTimestamp != nil && time.Now().After(candidate.DeletionTimestamp.Add(TerminatingPodTimeout)) {
+				return false, nil
+			}
+			return true, nil
+		},
+	},
+	{
+		// In Yellow or Red status only allow unhealthy Pods to be restarted.
+		// This is intended to unlock some situations where the cluster is not green and
+		// a Pod has to be restarted a second time.
+		name: "do_not_restart_healthy_node_if_not_green",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			green, err := context.esState.GreenHealth()
+			if err != nil {
+				return false, err
+			}
+			if green {
+				return true, nil
+			}
+			if !k8s.IsPodReady(candidate) {
+				return true, nil
+			}
+			return false, nil
+		},
+	},
+	{
+		// One master at a time
+		name: "one_master_at_a_time",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			// If candidate is not a master then we don't care
+			if !label.IsMasterNode(candidate) {
+				return true, nil
+			}
+			// If candidate is not healthy we want give it a chance to restart
+			_, healthy := context.healthyPods[candidate.Name]
+			if !healthy {
+				return true, nil
+			}
+
+			for _, pod := range deletedPods {
+				if label.IsMasterNode(pod) {
+					return false, nil
+				}
+			}
+			// Get the expected masters
+			expectedMasters := len(context.masterNodesNames)
+			// Get the healthy masters
+			healthyMasters := 0
+			for _, pod := range context.healthyPods {
+				if label.IsMasterNode(pod) {
+					healthyMasters++
+				}
+			}
+			return healthyMasters == expectedMasters, nil
+		},
+	},
+	{
+		// Force an upgrade of all the data nodes before upgrading the last master
+		name: "do_not_delete_last_master_if_datanodes_are_not_upgraded",
+		fn: func(
+			context PredicateContext,
+			candidate corev1.Pod,
+			deletedPods []corev1.Pod,
+			maxUnavailableReached bool,
+		) (b bool, e error) {
+			// If candidate is not a master then we don't care
+			if !label.IsMasterNode(candidate) {
+				return true, nil
+			}
+			for _, pod := range context.toUpdate {
+				if candidate.Name == pod.Name {
+					continue
+				}
+				if label.IsMasterNode(pod) {
+					// There are some other masters alive, allow this one to be deleted
+					return true, nil
+				}
+			}
+			// This is the last master, check if all data nodes are up to date
+			for _, pod := range context.toUpdate {
+				if label.IsDataNode(pod) {
```

Comment:
What if it is a mixed master + data node? You are potentially including the `candidate` here, could this lead to a situation where the last master+data node can never be upgraded?

Reply:
Good catch ! I'll fix that.